Include/internal/pycore_abstract.h

```diff
@@ -8,6 +8,21 @@ extern "C" {
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
+// Return 1 if 0 <= index < limit
+static inline int _Py_is_valid_index(Py_ssize_t index, Py_ssize_t limit)
+{
+    /* The cast to size_t lets us use just a single comparison
+        to check whether i is in the range: 0 <= i < limit.
+        See:  Section 14.2 "Bounds Checking" in the Agner Fog
+        optimization manual found at:
+        https://www.agner.org/optimize/optimizing_cpp.pdf
+
+        The function is not affected by -fwrapv, -fno-wrapv and -ftrapv
+        compiler options of GCC and clang
+    */
+    return (size_t)index < (size_t)limit;
```

Comment:
Should we `assert(index >= 0); assert(limit >= 0);`?

Reply:
@arhadthedev We should indeed `assert(limit >= 0);`. The check is not valid for negative values of `limit`.
The `index` is allowed to be negative, in that case this method should return 0